### ISSUE TITLE
Allow overrides of the default base OCI spec 

### ISSUE THREAD

**Issue by arnaldo2792** [moby/moby#45819]
### Description

In containerd, there is a configuration that can be set to override the base OCI spec used for the containers launched with the `CRI` plugin. In the [Bottlerocket OS], this feature is used to add/remove capabilities and certain RLIMITS, so that users don't have to add the same set of restrictions to all their container definitions. We are currently working on a patch for Docker to support the same use case in the Bottlerocket image build for ECS. Is this feature (override the base OCI spec) something that the Moby community is interested in supporting on the long run? If so, we are happy to contribute the patch (once ready) back to the moby project.

[Bottlerocket OS]: https://github.com/bottlerocket-os/bottlerocket

**Comment by neersighted** [comment]
I'm not sure overriding the raw OCI spec is something we're interested in upstream (e.g. if you need Moby with specialized adjustments, maintaining your own patches in Bottlerocket OS) might be more idiomatic. That being said, I think a design proposal or PR would certainly help decide if this is something that makes sense.

There are other tools we have recently added that allow similar (but not "default") changes as well, e.g. CDI, and it's also worth considering if a custom containerd shim/runtime or similar makes sense for your use-case (e.g. some of the things sysbox does).

cc @thaJeztah @AkihiroSuda @cpuguy83 @corhere

**Comment by thaJeztah** [comment]
Yes, this requires some discussion. I know there's some tickets related to "setting defaults" for containers (and we have some options in the daemon that already allow setting defaults, but they could use some looking at, as it's not always clear what options are "defaults for containers", which options are "baked" into the container (once created), etc).

Directly patching the OCI config feels like it may be a bit beyond scope (but could made easier / handed off with the CDI integration I guess).

All of the above said, prior discussions started to evolve around expected behavior, and "where" defaults should be set. When set on the daemon side, containers may run very "different" depending on how the daemon is configured, which for _some_ options could be reasonable, but may be a sliding slope.

From that perspective there's also been discussions to have a client-side config file that allows setting defaults.

**Comment by corhere** [comment]
> In containerd, there is a configuration that can be set to override the base OCI spec used for the containers launched with the `CRI` plugin.

<details><summary> Interesting, TIL! </summary>
<p>

https://github.com/containerd/containerd/blob/main/docs/cri/config.md
```toml
      # base_runtime_spec is a file path to a JSON file with the OCI spec that will be used as the base spec that all
      # container's are created from.
      # Use containerd's `ctr oci spec > /etc/containerd/cri-base.json` to output initial spec file.
      # Spec files are loaded at launch, so containerd daemon must be restarted on any changes to refresh default specs.
      # Still running containers and restarted containers will still be using the original spec from which that container was created.
      base_runtime_spec = ""
```

</p>
</details> 

While also related to OCI specs, that cri-containerd feature occupies a different niche than CDI and [NRI](https://github.com/containerd/nri). CDI and NRI patch the spec after the user configuration has been applied. CDI is opt-in per container only. And while an NRI plugin could implement whatever policy it wants on which container specs to patch, it would not have enough information to be able to implement a policy which emulates using a custom base spec as it won't know for sure which parts of the container spec were set explicitly versus inherited from the base spec.

Given both the existence of that cri-containerd option and of this issue, there clearly are use cases for (cloud) operators and other system administrators which are served by customizing the base spec used for all containers on a particular host. I think we should take this proposal seriously as it gives us a say in how it is implemented. In particular, we can insist on including in the `docker info` output the fact that a custom base spec has been configured on the daemon so that we will be aware of it in future bug reports. Otherwise, the Bottlerocket OS folks will ship docker packages built from sources patched with a different `oci.DefaultSpec()` or whatever than upstream anyway and we'd be none the wiser until some user files a bug report which we can't reproduce on a vanilla Moby build.

**Comment by vyaghras** [comment]
1. Changes in RLIMIT will work fine if we can add the ability for a setting in bottlerocket to end up essentially in the docker daemon.json .
2. For Capabilities patching can be done as follows
    1. Add the capabilities parameter in https://github.com/moby/moby/blob/dd61302f5bd87c887ec4cba22479a37ef78914c8/daemon/config/config_linux.go#L59 like `Capabilities` json:"capabilities,omitempty"` in linux config.
    2. Read the flag https://github.com/moby/moby/blob/dd61302f5bd87c887ec4cba22479a37ef78914c8/cmd/dockerd/config_unix.go#L30 
    3.  Adding the capabilities provided from etc/docker/daemon.json in https://github.com/moby/moby/blob/dd61302f5bd87c887ec4cba22479a37ef78914c8/daemon/oci_linux.go#L162 . For this we will keep the Capabilities provided by host configuration (container one) on priority. We will just append the daemon capabilities with default capabilities and make changes using TweakCapabilities function to change according to cap add and cap drop.

**Comment by arnaldo2792** [comment]
> Given both the existence of that cri-containerd option and of this issue, there clearly are use cases for (cloud) operators and other system administrators which are served by customizing the base spec used for all containers on a particular host. I think we should take this proposal seriously as it gives us a say in how it is implemented.

Cool! We will start working on our patch, and after it is approved by the Bottlerocket team we will begin contributing it back! Thank you all for the quick response. 

On a side note, I saw a few other GitHub issues in the repo that may be solved by adding support for this. I'll reference them here once I find them again.

**Comment by arnaldo2792** [comment]
Hello folks, while working on the implementation for this feature, we were considering if we should mimic what the CRI plugin does, and allow setting `BaseRuntimeSpec` per runtime; or just allow the one `BaseRuntimeSpec` value set at the daemon level and use that for all the runtimes.

Example configurations for each case are as follows:

- `BaseRuntimeSpec` per runtime:

```json
{
  "runtimes": { "my-cool-runtime": { "path": "runc", "base-runtime-spec": "path-to-file" } }
}
```

- `BaseRuntimeSpec` for all the runtimes

```json
{
  "base-runtime-spec": "/path/to/file"
}
```

Do you folks have a preference between the two? In our use case, the second option meets our needs but I will like to know your thoughts.

**Comment by corhere** [comment]
How about both? I'm sure cri-dockerd users would be thrilled to have more feature parity with cri-containerd. The per-runtime base spec configuration could override the daemon-wide config for that runtime.
